### PR TITLE
fix(48034): Get a literal string of type intersection in a template literal type

### LINKED ISSUE #48034: Template string type infers error when string intersect with object type
# Bug Report

### 🔎 Search Terms

Template string, Intersection type

### 🕗 Version & Regression Information

4.5.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=9&ssc=1&pln=9&pc=2#code/MYewdgzgLgBAhjAvDA5HFAoDUCeAHAUxgEEkZdCQAzeLCogWQEsAPM0gMhgG8qQQAXDGgAnJmADmAXzr4iAeTETxARjIADACTdiU9TAD0B8nJhMIqdNlOKmysACYN25iz2Hj9MxdHiJGGCA)

### 💻 Code

```ts
const a = 'a'

type A = typeof a

type Mix = A & {foo: string}

type Origin1 = `${A}` // type is 'a'
type Origin2 = `${Mix}` // type is string
```

### 🙁 Actual behavior

`Origin1` is `'a'` and `Origin2` is `string`

### 🙂 Expected behavior

`Origin1` and `Origin2` both are `'a'`

### PR DESCRIPTION
**Pull request by islandryu**
Fixes #48034

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -15134,12 +15134,31 @@ namespace ts {
                         newTexts.push(text);
                         text = texts[i + 1];
                     }
+                    else if (t.flags & TypeFlags.Intersection){
```
Comment (jakebailey) [inline src/compiler/checker.ts:15137]: Silly nit

```suggestion
                    else if (t.flags & TypeFlags.Intersection) {
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -15134,12 +15134,31 @@ namespace ts {
...
+
+            function getTemplateStringForInterSectionType(type: IntersectionType): string | undefined {
+                for (const t of type.types) {
+                    if (t.flags & (TypeFlags.Literal | TypeFlags.Null | TypeFlags.Undefined)) {
+                        return getTemplateStringForType(t) || "";
+                    }
+                    else if (t.flags & TypeFlags.TemplateLiteral) {
+                        if (!addSpans((t as TemplateLiteralType).texts, (t as TemplateLiteralType).types)) continue;
+                        return (t as TemplateLiteralType).texts[0];
+                    }
```
Comment (jakebailey) [inline src/compiler/checker.ts:15158]: This body looks suspiciously similar to the code above in `addSpans`; are they similar enough that they could all be pulled out into a single shared function?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -15117,25 +15117,38 @@ namespace ts {
...
+                        if (!addText) return false;
                     }
                     else {
-                        return false;
+                        if (isTextsArray) {
+                            return false;
+                        }
+                        else {
+                            continue;
+                        }
```
Comment (jakebailey) [inline src/compiler/checker.ts:15151]: This block can probably be:

```ts
else if (isTextsArray) {
    return false;
}
```

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -15117,25 +15117,38 @@ namespace ts {
...
+                        if (!isTextsArray) return true;
                     }
                     else if (isGenericIndexType(t) || isPatternLiteralPlaceholderType(t)) {
                         newTypes.push(t);
                         newTexts.push(text);
-                        text = texts[i + 1];
+                        text = addText;
+                    }
+                    else if (t.flags & TypeFlags.Intersection) {
+                        const addText = addSpans(texts[i + 1], (t as IntersectionType).types);
```
Comment (jakebailey) [inline src/compiler/checker.ts:15142]: This variable shadows the one above; maybe rename it to `added` or just write this as a plain if statement like other `addSpans` usages?

**Review by jakebailey — COMMENTED** [review]
Thanks for the refactor; seems better.

File: src/compiler/checker.ts (hunk 3 above)
Comment (islandryu) [reply src/compiler/checker.ts]: Fixed!

File: src/compiler/checker.ts (hunk 4 above)
Comment (islandryu) [reply src/compiler/checker.ts:15142]: Fixed!

**Review by jakebailey — APPROVED** [review]

**Pull request merged**